README.md
# emoji_data.rb

[![Gem Version](http://img.shields.io/gem/v/emoji_data.svg?style=flat)](https://rubygems.org/gems/emoji_data)
[![Build Status](http://img.shields.io/travis/mroth/emoji_data.rb.svg?style=flat)](https://travis-ci.org/mroth/emoji_data.rb)
[![Dependency Status](http://img.shields.io/gemnasium/mroth/emoji_data.rb.svg?style=flat)](https://gemnasium.com/mroth/emoji_data.rb)
[![Coverage Status](http://img.shields.io/coveralls/mroth/emoji_data.rb.svg?style=flat)](https://coveralls.io/r/mroth/emoji_data.rb)

Ruby library providing low level operations for dealing with Emoji
glyphs in the Unicode standard. :cool:

EmojiData is like a swiss-army knife for dealing with Emoji encoding issues. If
all you need to do is translate `:poop:` into :poop:, then there are plenty of
other libs out there that will probably do what you want.  But once you are
dealing with Emoji as a fundamental part of your application, and you start to
realize the nightmare of [doublebyte encoding][doublebyte] or
[variants][variant], then this library may be your new best friend.
:raised_hands:

EmojiData is used in production by [Emojitracker.com][emojitracker] to parse
well over 100M+ emoji glyphs daily. :dizzy:

Don't like Ruby? This library has also now been ported to [NodeJS][node] and
[Elixir][ex].

[doublebyte]: http://qr.ae/UBjMp
[variant]: http://www.unicode.org/L2/L2011/11438-emoji-var.pdf
[emojitracker]: http://www.emojitracker.com
[node]: https://github.com/mroth/emoji-data-js
[ex]: https://github.com/mroth/exmoji

## Installation

Add this line to your application's `Gemfile`:

    gem 'emoji_data'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install emoji_data

Currently requires `RUBY_VERSION >= 1.9.3`.

## Usage

### Documentation
Full API documentation is available via YARD or here:
http://rubydoc.info/github/mroth/emoji_data.rb/master/frames

### Examples
Here are some examples of the type of stuff you can do:

```irb
>> require 'emoji_data'
=> true

>> EmojiData.from_unified('1f680')
=> #<EmojiData::EmojiChar:0x007f8fdba33b40 @variations=[], @name="ROCKET",
@unified="1F680", @docomo=nil, @au="E5C8", @softbank="E10D", @google="FE7ED",
@image="1f680.png", @sheet_x=25, @sheet_y=4, @short_name="rocket",
@short_names=["rocket"], @text=nil, @apple_img=true, @hangouts_img=true,
@twitter_img=true>

>> EmojiData.all.count
=> 845

>> EmojiData.all_with_variants.count
=> 107

>> EmojiData.find_by_short_name("moon").count
=> 13

>> EmojiData.all.select(&:doublebyte?).map(&:short_name)
=> ["hash", "zero", "one", "two", "three", "four", "five", "six", "seven",
"eight", "nine", "cn", "de", "es", "fr", "gb", "it", "jp", "kr", "ru", "us"]

>> EmojiData.find_by_name("tree").map { |c| [c.unified, c.name, c.render] }
=> [["1F332", "EVERGREEN TREE", "🌲"], ["1F333", "DECIDUOUS TREE", "🌳"],
["1F334", "PALM TREE", "🌴"], ["1F384", "CHRISTMAS TREE", "🎄"], ["1F38B",
"TANABATA TREE", "🎋"]]

>> EmojiData.scan("I ♥ when marketers talk about the ☁. #blessed").each do |ec|
?>   puts "Found some #{ec.short_name}!"
>> end
Found some hearts!
Found some cloud!
=> [...]
```

## Contributing

Please be sure to run `rake spec` and help keep test coverage at :100:.

There is a full benchmark suite available via `scripts/benchmark.rb`.  Please
run before and after your changes to ensure you have not caused a performance
regression.

## License

[The MIT License (MIT)](LICENSE)


CHANGELOG.md
# Changelog

## 0.2.0 (5 September 2014)

 * Rename a number of methods to be clearer and more consistent with that they
   actually do:
     - `EmojiChar.char()` → `EmojiChar.render()`
     - `EmojiData.find_by_unified()` → `EmojiData.from_unified()`
     - `EmojiData.find_by_str()` → `EmojiData.scan()`

   Don't worry, the old names are still aliased in so you don't have to change
   anything in your existing code.  This change is make things clearer for
   people new to the library.

 * Add new `.from_short_name()` library method for fast keyword lookups.
 * DEVELOPERS: Internal code cleanup and better comments.
 * DEVELOPERS: Add benchmark suite for comparing method implementation time
   across versions of this library.

## 0.1.0 (3 May 2014)

 * Add support for Unicode variant encodings, used by MacOSX 10.9 / iOS 7.
   - For more info: http://www.unicode.org/L2/L2011/11438-emoji-var.pdf
   - By default, `EmojiChar.to_s()` and `.char()` will now use the variant
     encoding.
 * With adding support for variants, the speed of `find_by_str` regressed by
   approximately 20% (because there are more codepoints to match against). To
   counter this, we switched to a Regex based scan than improves performance of
   the method by over 250x(!).  A complete sorted search against 1000 strings
   now takes ~2ms where before it would take around a half second.
 * Import latest version of iamcal/emoji-data.
 * 100% test coverage. :sunglasses:

## 0.0.3 (1 April 2014)

 * On initialization, create hashmaps to cache lookups for `.find_by_unified()`.

   In a quick benchmark in MRI 2.1.1, this reduces the time needed for one
   million lookups from `13.5s` to `0.3s`!

   This is only for lookup by unified ID for now, since the other `find_by_*()`
   methods are actually searches that can return multiple values.  I'll look at
   nested hashmaps for those if there is a pressing performance need later.

## 0.0.2 (3 December 2013)

 * Remove JSON gem dependency since no longer supporting Ruby 1.8.7 anyhow.
 * Add `EmojiData.find_by_str` convenience method to match on a string.
 * Make default `EmojiChar.to_s()` the same as `EmojiChar.char()`


## 0.0.1 (7 June 2013)

 * Initial release


spec/spec_helper.rb
require 'rspec'
require 'simplecov'
require 'coveralls'
SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter[
  SimpleCov::Formatter::HTMLFormatter,
  Coveralls::SimpleCov::Formatter
]
SimpleCov.start do
  add_filter '/spec'
  add_filter '/.bundle'
end

require 'emoji_data'
include EmojiData

RSpec.configure do |config|
  config.color_enabled = true
  config.formatter     = 'documentation'
end


spec/emoji_char_spec.rb
#encoding: UTF-8
require 'spec_helper'

describe EmojiChar do
  describe ".new" do
    before(:all) do
      poop_json = %q/{"name":"PILE OF POO","unified":"1F4A9","variations":[],"docomo":"","au":"E4F5","softbank":"E05A","google":"FE4F4","image":"1f4a9.png","sheet_x":11,"sheet_y":19,"short_name":"hankey","short_names":["hankey","poop","shit"],"text":null}/
      @poop = EmojiChar.new(JSON.parse poop_json)
    end
    it "should create instance getters for all key-values in emoji.json, with blanks as nil" do
      @poop.name.should eq('PILE OF POO')
      @poop.unified.should eq('1F4A9')
      @poop.variations.should eq([])
      @poop.docomo.should eq('')
      @poop.au.should eq('E4F5')
      @poop.softbank.should eq('E05A')
      @poop.google.should eq('FE4F4')
      @poop.image.should eq('1f4a9.png')
      @poop.sheet_x.should eq(11)
      @poop.sheet_y.should eq(19)
      @poop.short_name.should eq('hankey')
      @poop.short_names.should eq(["hankey","poop","shit"])
      @poop.text.should eq(nil)
    end
  end

  context "instance methods" do
    before(:all) do
      @invader   = EmojiChar.new({'unified' => '1F47E'})
      @usflag    = EmojiChar.new({'unified' => '1F1FA-1F1F8'})
      @hourglass = EmojiChar.new({'unified' => '231B', 'variations' => ['231B-FE0F']})
      @cloud     = EmojiChar.new({'unified' => '2601', 'variations' => ['2601-FE0F']})
    end

    describe "#to_s" do
      it "should return the unicode char as string as default to_s" do
        @invader.to_s.should eq(@invader.char)
      end
    end

    describe "#render" do
      it "should render as happy shiny unicode" do
        @invader.render.should eq("👾")
      end
      it "should render as happy shiny unicode for doublebyte chars too" do
        @usflag.render.should eq("🇺🇸")
      end
      it "should have a flag to output forced emoji variant char encoding if requested" do
        @cloud.render(    {variant_encoding: false}).should eq("\u{2601}")
        @cloud.render(    {variant_encoding:  true}).should eq("\u{2601}\u{FE0F}")
        @invader.render(  {variant_encoding: false}).should eq("\u{1F47E}")
        @invader.render(  {variant_encoding:  true}).should eq("\u{1F47E}")
      end
      it "should default to variant encoding for chars with a variant present" do
        @cloud.render.should eq("\u{2601}\u{FE0F}")
        @hourglass.render.should eq("\u{231B}\u{FE0F}")
      end
    end

    describe "#char - DEPRECATED" do
      it "should maintain compatibility with old method name for .render" do
        @cloud.char.should eq(@cloud.render)
      end
    end

    describe "#chars" do
      it "should return an array of all possible string render variations" do
        @invader.chars.should eq(["\u{1F47E}"])
        @cloud.chars.should   eq(["\u{2601}","\u{2601}\u{FE0F}"])
      end
    end

    describe "#doublebyte?" do
      it "should indicate when a character is doublebyte based on the unified ID" do
        @usflag.doublebyte?.should be_true
        @invader.doublebyte?.should be_false
      end
    end

    describe "#variant?" do
      it "should indicate when a character has an alternate variant encoding" do
        @hourglass.variant?.should be_true
        @usflag.variant?.should be_false
      end
    end

    describe "#variant" do
      it "should return the most likely variant encoding ID representation for the char" do
        @hourglass.variant.should eq('231B-FE0F')
      end
      it "should return nil if no variant encoding for the char exists" do
        @usflag.variant.should be_nil
      end
    end
  end
end


spec/emoji_data_spec.rb
#encoding: UTF-8
require 'spec_helper'

describe EmojiData do
  describe ".all" do
    it "should return an array of all 845 known emoji chars" do
      EmojiData.all.count.should eq(845)
    end
    it "should return all EmojiChar objects" do
      EmojiData.all.all? {|char| char.class == EmojiData::EmojiChar}.should be_true
    end
  end

  describe ".all_doublebyte" do
    it "should return an array of all 21 known emoji chars with doublebyte encoding" do
      EmojiData.all_doublebyte.count.should eq(21)
    end
  end

  describe ".all_with_variants" do
    it "should return an array of all 107 known emoji chars with variant encodings" do
      EmojiData.all_with_variants.count.should eq(107)
    end
  end

  describe ".chars" do
    it "should return an array of all chars in unicode string format" do
      EmojiData.chars.all? {|char| char.class == String}.should be_true
    end
    it "should by default return one entry per known EmojiChar" do
      EmojiData.chars.count.should eq(EmojiData.all.count)
    end
    it "should include variants in list when options {include_variants: true}" do
      results = EmojiData.chars({include_variants: true})
      numChars    = EmojiData.all.count
      numVariants = EmojiData.all_with_variants.count
      results.count.should eq(numChars + numVariants)
    end
    it "should not have any duplicates in list when variants are included" do
      results = EmojiData.chars({include_variants: true})
      results.count.should eq(results.uniq.count)
    end
  end

  describe ".codepoints" do
    it "should return an array of all known codepoints in dashed string representation" do
      EmojiData.codepoints.all? {|cp| cp.class == String}.should be_true
      EmojiData.codepoints.all? {|cp| cp.match(/^[0-9A-F\-]{4,11}$/)}.should be_true
    end
    it "should include variants in list when options {include_variants: true}" do
      results = EmojiData.codepoints({include_variants: true})
      numChars    = EmojiData.all.count
      numVariants = EmojiData.all_with_variants.count
      results.count.should eq(numChars + numVariants)
      results.all? {|cp| cp.match(/^[0-9A-F\-]{4,16}$/)}.should be_true
    end
  end

  describe ".scan" do
    before(:all) do
      @exact_results   = EmojiData.scan("🚀")
      @multi_results   = EmojiData.scan("flying on my 🚀 to visit the 👾 people.")
      @variant_results = EmojiData.scan("\u{0023}\u{FE0F}\u{20E3}")
      @variant_multi   = EmojiData.scan("first a \u{0023}\u{FE0F}\u{20E3} then a 🚀")
    end
    it "should find the proper EmojiChar object from a single string char" do
      @exact_results.should be_kind_of(Array)
      @exact_results.length.should eq(1)
      @exact_results.first.should be_kind_of(EmojiChar)
      @exact_results.first.name.should eq('ROCKET')
    end
    it "should find the proper EmojiChar object from a variant encoded char" do
      @variant_results.length.should eq(1)
      @variant_results.first.name.should eq('HASH KEY')
    end
    it "should match multiple chars from within a string" do
      @multi_results.should be_kind_of(Array)
      @multi_results.length.should eq(2)
      @multi_results[0].should be_kind_of(EmojiChar)
      @multi_results[1].should be_kind_of(EmojiChar)
    end
    it "should return multiple matches in the proper order" do
      @multi_results[0].name.should eq('ROCKET')
      @multi_results[1].name.should eq('ALIEN MONSTER')
    end
    it "should return multiple matches in the proper order for variant encodings" do
      @variant_multi[0].name.should eq('HASH KEY')
      @variant_multi[1].name.should eq('ROCKET')
    end
  end

  describe ".find_by_str - DEPRECATED" do
    it "should maintain compatibility with old method name for .scan" do
      EmojiData.find_by_str("\u{0023}\u{FE0F}\u{20E3}").should eq(EmojiData.scan("\u{0023}\u{FE0F}\u{20E3}"))
    end
  end

  describe ".from_unified" do
    it "should find the proper EmojiChar object" do
      results = EmojiData.from_unified('1f680')
      results.should be_kind_of(EmojiChar)
      results.name.should eq('ROCKET')
    end
    it "should normalise capitalization for hex values" do
      EmojiData.from_unified('1f680').should_not be_nil
    end
    it "should find via variant encoding ID format as well" do
      results = EmojiData.from_unified('2764-fe0f')
      results.should_not be_nil
      results.name.should eq('HEAVY BLACK HEART')
    end
  end

  describe ".find_by_unified - DEPRECATED" do
    it "should maintain compatibility with old method name for .from_unified" do
      EmojiData.find_by_unified('1f680').should eq(EmojiData.from_unified('1f680'))
    end
  end

  describe ".find_by_name" do
    it "returns an array of results, upcasing input if needed" do
      EmojiData.find_by_name('tree').should be_kind_of(Array)
      EmojiData.find_by_name('tree').count.should eq(5)
    end
    it "returns [] if nothing is found" do
      EmojiData.find_by_name('sdlkfjlskdfj').should_not be_nil
      EmojiData.find_by_name('sdlkfjlskdfj').should be_kind_of(Array)
      EmojiData.find_by_name('sdlkfjlskdfj').count.should eq(0)
    end
  end

  describe ".find_by_short_name" do
    it "returns an array of results, downcasing input if needed" do
      EmojiData.find_by_short_name('MOON').should be_kind_of(Array)
      EmojiData.find_by_short_name('MOON').count.should eq(13)
    end
    it "returns [] if nothing is found" do
      EmojiData.find_by_short_name('sdlkfjlskdfj').should_not be_nil
      EmojiData.find_by_short_name('sdlkfjlskdfj').should be_kind_of(Array)
      EmojiData.find_by_short_name('sdlkfjlskdfj').count.should eq(0)
    end
  end

  describe ".from_short_name" do
    it "returns exact matches on a short name" do
      results = EmojiData.from_short_name('scream')
      results.should be_kind_of(EmojiChar)
      results.name.should eq('FACE SCREAMING IN FEAR')
    end
    it "handles lowercasing input if required" do
      EmojiData.from_short_name('SCREAM').should eq( EmojiData.from_short_name('scream') )
    end
    it "works on secondary keywords" do
      primary = EmojiData.from_short_name('hankey')
      EmojiData.from_short_name('poop').should eq(primary)
      EmojiData.from_short_name('shit').should eq(primary)
    end
    it "returns nil if nothing matches" do
      EmojiData.from_short_name('taco').should be_nil
    end
  end

  describe ".char_to_unified" do
    it "converts normal emoji to unified codepoint" do
      EmojiData.char_to_unified("👾").should eq('1F47E')
      EmojiData.char_to_unified("🚀").should eq('1F680')
    end
    it "converts double-byte emoji to proper codepoint" do
      EmojiData.char_to_unified("🇺🇸").should eq('1F1FA-1F1F8')
      EmojiData.char_to_unified("#⃣").should eq('0023-20E3')
    end
    it "converts variant encoded emoji to variant unified codepoint" do
      EmojiData.char_to_unified("\u{2601}\u{FE0F}").should eq('2601-FE0F')
    end
  end

  # TODO: below is kinda redundant but it is helpful as a helper method so maybe still test
  describe ".unified_to_char" do
    it "converts normal unified codepoints to unicode strings" do
      EmojiData.unified_to_char('1F47E').should eq("👾")
      EmojiData.unified_to_char('1F680').should eq("🚀")
    end
    it "converts doublebyte unified codepoints to unicode strings" do
      EmojiData.unified_to_char('1F1FA-1F1F8').should eq("🇺🇸")
      EmojiData.unified_to_char('0023-20E3').should eq("#⃣")
    end
    it "converts variant unified codepoints to unicode strings" do
      EmojiData.unified_to_char('2764-fe0f').should eq("\u{2764}\u{FE0F}")
    end
    it "converts variant+doublebyte chars (triplets!) to unicode strings" do
      EmojiData.unified_to_char('0030-FE0F-20E3').should eq("\u{0030}\u{FE0F}\u{20E3}")
    end
  end
end


bench.rb
require './lib/emoji_data'
require 'benchmark/ips'

suites = []

s0 = "I liek to eat cake oh so very much cake eating is nice!! #cake #food"
s1 = "🚀"
s2 = "flying on my 🚀 to visit the 👾 people."
s3 = "first a \u{0023}\u{FE0F}\u{20E3} then a 🚀"

suites << Benchmark.ips do |x|
  x.config(:time => 1, :warmup => 0)
  x.report("EmojiData.scan(s0)") { EmojiData.scan(s0) }
  x.report("EmojiData.scan(s1)") { EmojiData.scan(s1) }
  x.report("EmojiData.scan(s2)") { EmojiData.scan(s2) }
  x.report("EmojiData.scan(s3)") { EmojiData.scan(s3) }
end


suites << Benchmark.ips do |x|
  x.config(:time => 1, :warmup => 0)
  x.report("EmojiData.all")                       { EmojiData.all() }
  x.report("EmojiData.all_doublebyte")            { EmojiData.all_doublebyte() }
  x.report("EmojiData.all_with_variants")         { EmojiData.all_with_variants() }
  x.report("EmojiData.from_unified")              { EmojiData.from_unified("1F680") }
  x.report("EmojiData.chars")                     { EmojiData.chars() }
  x.report("EmojiData.codepoints")                { EmojiData.codepoints() }
  x.report("EmojiData.find_by_name - many")       { EmojiData.find_by_name("tree") }
  x.report("EmojiData.find_by_name - none")       { EmojiData.find_by_name("zzzz") }
  x.report("EmojiData.find_by_short_name - many") { EmojiData.find_by_short_name("MOON") }
  x.report("EmojiData.find_by_short_name - none") { EmojiData.find_by_short_name("zzzz") }
  x.report("EmojiData.char_to_unified - single")  { EmojiData.char_to_unified("🚀") }
  x.report("EmojiData.char_to_unified - double")  { EmojiData.char_to_unified("\u{2601}\u{FE0F}") }
  x.report("EmojiData.unified_to_char - single")  { EmojiData.unified_to_char("1F47E") }
  x.report("EmojiData.unified_to_char - double")  { EmojiData.unified_to_char("2764-fe0f") }
  x.report("EmojiData.unified_to_char - triple")  { EmojiData.unified_to_char("0030-FE0F-20E3") }
end


invader   = EmojiData::EmojiChar.new({unified: '1F47E'})
usflag    = EmojiData::EmojiChar.new({unified: '1F1FA-1F1F8'})
hourglass = EmojiData::EmojiChar.new({unified: '231B', variations: ['231B-FE0F']})
cloud     = EmojiData::EmojiChar.new({unified: '2601', variations: ['2601-FE0F']})

suites << Benchmark.ips do |x|
  x.config(:time => 1, :warmup => 0)
  x.report("EmojiChar.render - single")  { invader.render() }
  x.report("EmojiChar.render - double")  { usflag.render() }
  x.report("EmojiChar.render - variant") { cloud.render({variant_encoding: true}) }
  x.report("EmojiChar.chars")            { cloud.chars() }
  x.report("EmojiChar.doublebyte?")      { invader.doublebyte?() }
  x.report("EmojiChar.variant?")         { invader.variant?() }
  x.report("EmojiChar.variant")          { invader.variant() }
end


def micros(hz)
  1_000_000 / hz
end

suites.each do |report|
  results = report.entries.sort { |a,b| b.ips <=> a.ips }

  print "\n"
  results.each do |r|
    printf "%-45s %10u   %.2f µs/op\n", r.label, r.iterations, micros(r.ips)
  end
end


lib/emoji_data.rb
require 'emoji_data/version'
require 'emoji_data/emoji_char'
require 'json'

module EmojiData

  # specify some location paths
  GEM_ROOT = File.join(File.dirname(__FILE__), '..')
  VENDOR_DATA = 'vendor/emoji-data/emoji.json'

  # precomputed list of all possible emoji characters
  EMOJI_CHARS = begin
    raw_json = IO.read(File.join(GEM_ROOT, VENDOR_DATA))
    vendordata = JSON.parse( raw_json )
    vendordata.map { |em| EmojiChar.new(em) }
  end

  # precomputed hashmap for fast precached lookups in .from_unified
  EMOJICHAR_UNIFIED_MAP = {}
  EMOJI_CHARS.each do |ec|
    EMOJICHAR_UNIFIED_MAP[ec.unified] = ec
    ec.variations.each  { |variant| EMOJICHAR_UNIFIED_MAP[variant] = ec }
  end

  # precomputed hashmap for fast precached lookups in .from_short_name
  EMOJICHAR_KEYWORD_MAP = {}
  EMOJI_CHARS.each do |ec|
    ec.short_names.each { |keyword| EMOJICHAR_KEYWORD_MAP[keyword] = ec }
  end

  # our constants are only for usage internally
  private_constant :GEM_ROOT, :VENDOR_DATA
  private_constant :EMOJI_CHARS, :EMOJICHAR_UNIFIED_MAP, :EMOJICHAR_KEYWORD_MAP


  # Returns a list of all known Emoji characters as `EmojiChar` objects.
  #
  # @return [Array<EmojiChar>] a list of all known `EmojiChar`.
  def self.all
    EMOJI_CHARS
  end

  # Returns a list of all `EmojiChar` that are represented with doublebyte
  # encoding.
  #
  # @return [Array<EmojiChar>] a list of all doublebyte `EmojiChar`.
  def self.all_doublebyte
    EMOJI_CHARS.select(&:doublebyte?)
  end

  # Returns a list of all `EmojiChar` that have at least one variant encoding.
  #
  # @return [Array<EmojiChar>] a list of all `EmojiChar` with variant encoding.
  def self.all_with_variants
    EMOJI_CHARS.select(&:variant?)
  end

  # Returns a list of all known Emoji characters rendered as UTF-8 strings.
  #
  # By default, the default rendering options for this library will be used.
  # However, if you pass an option hash with `include_variants: true` then all
  # possible renderings of a single glyph will be included, meaning that:
  #
  # 1. You will have "duplicate" emojis in your list.
  # 2. This list is now suitable for exhaustably matching against in a search.
  #
  # @option opts [Boolean] :include_variants whether or not to include all
  #   possible encoding variants in the list
  #
  # @return [Array<String>] all Emoji characters rendered as UTF-8 strings
  def self.chars(opts={})
    options = {include_variants: false}.merge(opts)

    normals = EMOJI_CHARS.map { |c| c.render({variant_encoding: false}) }

    if options[:include_variants]
      extras  = self.all_with_variants.map { |c| c.render({variant_encoding: true}) }
      return normals + extras
    end
    normals
  end

  # Returns a list of all known codepoints representing Emoji characters.
  #
  # @option (see .chars)
  # @return [Array<String>] all codepoints represented as unified ID strings
  def self.codepoints(opts={})
    options = {include_variants: false}.merge(opts)

    normals = EMOJI_CHARS.map(&:unified)

    if options[:include_variants]
      extras = self.all_with_variants.map {|c| c.variant}
      return normals + extras
    end
    normals
  end

  # Convert a native UTF-8 string glyph to its unified codepoint ID.
  #
  # This is a conversion operation, not a match, so it may produce unexpected
  # results with different types of values.
  #
  # @param char [String] a single rendered emoji glyph encoded as a UTF-8 string
  # @return [String] the unified ID
  #
  # @example
  #   >> EmojiData.unified_to_char("1F47E")
  #   => "👾"
  def self.char_to_unified(char)
    char.codepoints.to_a.map { |i| i.to_s(16).rjust(4,'0')}.join('-').upcase
  end

  # Convert a unified codepoint ID directly to its UTF-8 string representation.
  #
  # @param uid [String] the unified codepoint ID for an emoji
  # @return [String] UTF-8 string rendering of the emoji character
  #
  # @example
  #   >> EmojiData.char_to_unified("👾")
  #   => "1F47E"
  def self.unified_to_char(uid)
    EmojiChar::unified_to_char(uid)
  end

  # Finds a specific `EmojiChar` based on its unified codepoint ID.
  #
  # @param uid [String] the unified codepoint ID for an emoji
  # @return [EmojiChar]
  def self.from_unified(uid)
    EMOJICHAR_UNIFIED_MAP[uid.upcase]
  end

  # precompile regex pattern for fast matches in `.scan`
  # needs to be defined after self.chars so not at top of file for now...
  FBS_REGEXP = Regexp.new(
    "(?:#{EmojiData.chars({include_variants: true}).join("|")})"
  )
  private_constant :FBS_REGEXP

  # Scans a string for all encoded emoji characters contained within.
  #
  # @param str [String] the target string to search
  # @return [Array<EmojiChar>] all emoji characters contained within the target
  #    string, in the order they appeared.
  #
  # @example
  #   >> EmojiData.scan("flying on my 🚀 to visit the 👾 people.")
  #   => [#<EmojiData::EmojiChar... @name="ROCKET", @unified="1F680", ...>,
  #   #<EmojiData::EmojiChar... @name="ALIEN MONSTER", @unified="1F47E", ...>]
  def self.scan(str)
    matches = str.scan(FBS_REGEXP)
    matches.map { |m| EmojiData.from_unified(EmojiData.char_to_unified(m)) }
  end

  # Finds any `EmojiChar` that contains given string in its official name.
  #
  # @param name [String]
  # @return [Array<EmojiChar>]
  def self.find_by_name(name)
    self.find_by_value(:name, name.upcase)
  end

  # Find all `EmojiChar` that match string in any of their associated short
  # name keywords.
  #
  # @param short_name [String]
  # @return [Array<EmojiChar>]
  def self.find_by_short_name(short_name)
    self.find_by_value(:short_name, short_name.downcase)
  end

  # Finds a specific `EmojiChar` based on the unified codepoint ID.
  #
  # Must be exact match.
  #
  # @param short_name [String]
  # @return [EmojiChar]
  def self.from_short_name(short_name)
    EMOJICHAR_KEYWORD_MAP[short_name.downcase]
  end

  # alias old method names for legacy apps
  class << self
    alias_method :find_by_unified, :from_unified
    alias_method :find_by_str, :scan
  end


  protected

  def self.find_by_value(field,value)
    self.all.select { |char| char.send(field).include? value }
  end

end


lib/emoji_data/emoji_char.rb
module EmojiData

  # EmojiChar represents a single Emoji character and its associated metadata.
  #
  # @!attribute name
  #   @return [String] The standardized name used in the Unicode specification
  #     to represent this emoji character.
  #
  # @!attribute unified
  #   @return [String] The primary unified codepoint ID for the emoji character.
  #
  # @!attribute variations
  #   @return [Array<String>] A list of all variant codepoints that may also
  #     represent this emoji.
  #
  # @!attribute short_name
  #   @return [String] The canonical "short name" or keyword used in many
  #     systems to refer to this emoji. Often surrounded by `:colons:` in
  #     systems like GitHub & Campfire.
  #
  # @!attribute short_names
  #   @return [Array<String>] A full list of possible keywords for the emoji.
  #
  # @!attribute text
  #   @return [String] An alternate textual representation of the emoji, for
  #   example a smiley face emoji may be represented with an ASCII alternative.
  #   Most emoji do not have a text alternative. This is typically used when
  #   building an automatic translation from typed emoticons.
  #
  class EmojiChar

    def initialize(emoji_hash)
      # work around inconsistency in emoji.json for now by just setting a blank
      # array for instance value, and let it get overriden in main
      # deserialization loop if variable is present.
      @variations = []

      # trick for declaring instance variables while iterating over a hash
      # http://stackoverflow.com/questions/1615190/
      emoji_hash.each do |k,v|
        instance_variable_set("@#{k}",v)
        eigenclass = class<<self; self; end
        eigenclass.class_eval { attr_reader k }
      end
    end

    # Renders an `EmojiChar` to its string glyph representation, suitable for
    # printing to screen.
    #
    # @option opts [Boolean] :variant_encoding specify whether the variant
    #   encoding selector should be used to hint to rendering devices that
    #   "graphic" representation should be used. By default, we use this for all
    #   Emoji characters that contain a possible variant.
    #
    # @return [String] the emoji character rendered to a UTF-8 string
    def render(opts = {})
      options = {variant_encoding: true}.merge(opts)
      #decide whether to use the normal unified ID or the variant for encoding to str
      target = (self.variant? && options[:variant_encoding]) ? self.variant : @unified
      EmojiChar::unified_to_char(target)
    end

    alias_method :to_s, :render
    alias_method :char, :render

    # Returns a list of all possible UTF-8 string renderings of an `EmojiChar`.
    #
    # E.g., normal, with variant selectors, etc. This is useful if you want to
    # have all possible values to match against when searching for the emoji in
    # a string representation.
    #
    # @return [Array<String>] all possible UTF-8 string renderings
    def chars
      results = [self.render({variant_encoding: false})]
      @variations.each do |variation|
        results << EmojiChar::unified_to_char(variation)
      end
      @chars ||= results
    end

    # Is the `EmojiChar` represented by a doublebyte codepoint in Unicode?
    #
    # @return [Boolean]
    def doublebyte?
      @unified.include? "-"
    end

    # Does the `EmojiChar` have an alternate Unicode variant encoding?
    #
    # @return [Boolean]
    def variant?
      @variations.length > 0
    end

    # Returns the most likely variant-encoding codepoint ID for an `EmojiChar`.
    #
    # For now we only know of one possible variant encoding for certain
    # characters, but there could be others in the future.
    #
    # This is typically used to force Emoji rendering for characters that could
    # be represented in standard font glyphs on certain operating systems.
    #
    # The resulting encoded string will be two codepoints, or three codepoints
    # for doublebyte Emoji characters.
    #
    # @return [String, nil]
    #   The most likely variant-encoding codepoint ID.
    #   If there is no variant-encoding for a character, returns nil.
    def variant
      @variations.first
    end


    protected

    def self.unified_to_char(cps)
      cps.split('-').map { |i| i.hex }.pack("U*")
    end

  end
end


scripts/benchmark.rb
# encoding: UTF-8

require './lib/emoji_data'
require 'benchmark/ips'

suites = []

s0 = "I liek to eat cake oh so very much cake eating is nice!! #cake #food"
s1 = "🚀"
s2 = "flying on my 🚀 to visit the 👾 people."
s3 = "first a \u{0023}\u{FE0F}\u{20E3} then a 🚀"

suites << Benchmark.ips do |x|
  x.config(:time => 1, :warmup => 0)
  x.report("EmojiData.scan(s0)") { EmojiData.scan(s0) }
  x.report("EmojiData.scan(s1)") { EmojiData.scan(s1) }
  x.report("EmojiData.scan(s2)") { EmojiData.scan(s2) }
  x.report("EmojiData.scan(s3)") { EmojiData.scan(s3) }
end


suites << Benchmark.ips do |x|
  x.config(:time => 1, :warmup => 0)
  x.report("EmojiData.all")                       { EmojiData.all() }
  x.report("EmojiData.all_doublebyte")            { EmojiData.all_doublebyte() }
  x.report("EmojiData.all_with_variants")         { EmojiData.all_with_variants() }
  x.report("EmojiData.from_unified")              { EmojiData.from_unified("1F680") }
  x.report("EmojiData.chars")                     { EmojiData.chars() }
  x.report("EmojiData.codepoints")                { EmojiData.codepoints() }
  x.report("EmojiData.find_by_name - many")       { EmojiData.find_by_name("tree") }
  x.report("EmojiData.find_by_name - none")       { EmojiData.find_by_name("zzzz") }
  x.report("EmojiData.find_by_short_name - many") { EmojiData.find_by_short_name("MOON") }
  x.report("EmojiData.find_by_short_name - none") { EmojiData.find_by_short_name("zzzz") }
  x.report("EmojiData.char_to_unified - single")  { EmojiData.char_to_unified("🚀") }
  x.report("EmojiData.char_to_unified - double")  { EmojiData.char_to_unified("\u{2601}\u{FE0F}") }
  x.report("EmojiData.unified_to_char - single")  { EmojiData.unified_to_char("1F47E") }
  x.report("EmojiData.unified_to_char - double")  { EmojiData.unified_to_char("2764-fe0f") }
  x.report("EmojiData.unified_to_char - triple")  { EmojiData.unified_to_char("0030-FE0F-20E3") }
end


invader   = EmojiData::EmojiChar.new({unified: '1F47E'})
usflag    = EmojiData::EmojiChar.new({unified: '1F1FA-1F1F8'})
hourglass = EmojiData::EmojiChar.new({unified: '231B', variations: ['231B-FE0F']})
cloud     = EmojiData::EmojiChar.new({unified: '2601', variations: ['2601-FE0F']})

suites << Benchmark.ips do |x|
  x.config(:time => 1, :warmup => 0)
  x.report("EmojiChar.render - single")  { invader.render() }
  x.report("EmojiChar.render - double")  { usflag.render() }
  x.report("EmojiChar.render - variant") { cloud.render({variant_encoding: true}) }
  x.report("EmojiChar.chars")            { cloud.chars() }
  x.report("EmojiChar.doublebyte?")      { invader.doublebyte?() }
  x.report("EmojiChar.variant?")         { invader.variant?() }
  x.report("EmojiChar.variant")          { invader.variant() }
end


def micros(hz)
  1_000_000 / hz
end

suites.each do |report|
  results = report.entries.sort { |a,b| b.ips <=> a.ips }

  print "\n"
  results.each do |r|
    printf "%-45s %10u   %.2f µs/op\n", r.label, r.iterations, micros(r.ips)
  end
end
